tough/src/transport.rs

```diff
@@ -1,36 +1,172 @@
-use std::io::Read;
+#[cfg(feature = "http")]
+use crate::{ClientSettings, HttpTransport};
+use dyn_clone::DynClone;
+use snafu::Snafu;
+use std::fmt::Debug;
+use std::io::{ErrorKind, Read};
 use url::Url;
 
 /// A trait to abstract over the method/protocol by which files are obtained.
-pub trait Transport {
-    /// The type of `Read` object that the `fetch` function will return.
-    type Stream: Read;
+///
+/// The trait hides the underlying types involved by returning the `Read` object as a
+/// `Box<dyn Read + Send>` and by requiring concrete type [`TransportError`] as the error type.
+///
+pub trait Transport: Debug + DynClone {
+    /// Opens a `Read` object for the file specified by `url`.
+    fn fetch(&self, url: Url) -> Result<Box<dyn Read + Send>, TransportError>;
+}
 
-    /// The type of error that the `fetch` function will return.
-    type Error: std::error::Error + Send + Sync + 'static;
+// Implement `Clone` for `Transport` trait objects.
+dyn_clone::clone_trait_object!(Transport);
 
-    /// Opens a `Read` object for the file specified by `url`.
-    fn fetch(&self, url: Url) -> Result<Self::Stream, Self::Error>;
+// =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=   =^..^=
+
+/// The kind of error that the transport object experienced during `fetch`.
+///
+/// # Why
+///
+/// Some TUF operations need to know if a [`Transport`] failure is a result of a file not being
+/// found. In particular:
+/// > 5.1.2. Try downloading version N+1 of the root metadata file `[...]` If this file is not
+/// > available `[...]` then go to step 5.1.9.
+///
+/// To distinguish this case from other [`Transport`] failures, we use
+/// `TransportErrorKind::FileNotFound`.
+///
+#[derive(Debug, Copy, Clone)]
+#[non_exhaustive]
+pub enum TransportErrorKind {
+    /// The trait does not handle the URL scheme named in `String`. e.g. `file://` or `http://`.
+    UnsupportedUrlScheme,
+    /// The file cannot be found.
+    FileNotFound,
+    /// The transport failed for any other reason, e.g. IO error, HTTP broken pipe, etc.
+    Other,
+}
+
+/// The error type that [`Transport`] `fetch` returns.
+#[derive(Debug, Snafu)]
+#[snafu(visibility = "pub")]
```

Comment:
Re-exporting the generated partial error structs from snafu makes part of our API based on a not-yet-1.0 library. I think manually implementing `std::error::Error` is a better, more-standard experience for library developers implementing `Transport`.

Reply:
How about an [opaque struct](https://docs.rs/snafu/0.6.10/snafu/guide/opaque/index.html) that only implements a few traits and doesn't expose internals?

Reply:
Opaque structs might solve the separate issue of our error variants being particularly purpose-specific, which makes stabilization more difficult.

The main concern I had when seeing this visibility parameter was the partial `IntoError` structs being part of the tough library interface. One thing I didn't notice is that `mod transport` isn't `pub`, so the partial structs here don't end up in the interface. But just to be on the safe side...

```suggestion
#[snafu(visibility = "pub(crate)")]
```